vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java

```diff
@@ -90,9 +93,13 @@ public void close() {
     MongoCollection<JsonObject> coll = getCollection(collection, writeOption);
     Object id = document.getValue(ID_FIELD);
     if (id == null) {
-      coll.insertOne(document, convertCallback(resultHandler, wr -> document.getString(ID_FIELD)));
+      coll.insertOne(document, convertCallback(resultHandler, wr -> useObjectId ? document.getJsonObject(ID_FIELD).getString(OID_FIELD) : document.getString(ID_FIELD)));
     } else {
-      coll.replaceOne(wrap(new JsonObject().put(ID_FIELD, document.getValue(ID_FIELD))), document, convertCallback(resultHandler, result -> null));
+      JsonObject filter = new JsonObject();
+      encodeKeyWhenUseObjectId(document);
```

Comment:
Since these methods return the document again, I would do:
```
document=encodeKeyWhenUseObjectId(document);
```
To make it obvious that document is being modified.

If you can also do this on all uses of `decodeKeyWhenUseObjectId` and `encodeKeyWhenUseObjectId`